src/Tutorials/Custom-Grain-Storage.md

```diff
@@ -0,0 +1,232 @@
+
+## Initializing the storage
+
+To initialize the storage, we register an `Init` function on the `ApplicationServices` lifecycle.
+
+```csharp
+public void Participate(ISiloLifecycle lifecycle)
+{
+    lifecycle.Subscribe(OptionFormattingUtilities.Name<FileGrainStorage>(_storageName), ServiceLifecycleStage.ApplicationServices, Init);
+}
+```
+
+The `Init` function is used to set the `_jsonSettings` which will be used to configure the `Json` serializer. At the same time we create the folder to store the grains states if it does not exist yet.
+
+```csharp
+private Task Init(CancellationToken ct)
+{
+    // Settings could be made configurable from Options.
+    _jsonSettings = OrleansJsonSerializer.UpdateSerializerSettings(OrleansJsonSerializer.GetDefaultSerializerSettings(_typeResolver, _grainFactory), false, false, null);
+
+    var directory = new System.IO.DirectoryInfo(_rootDirectory);
+    if (!directory.Exists)
+        directory.Create();
+
+    return Task.CompletedTask;
+}
+```
+
+## Reading State
+
+To read the state, to ensure uniqueness, we construct the filename using the grain key and the grain type.
+
+```csharp
+public async Task ReadStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+{
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+    if (!fileInfo.Exists)
+        return;
+
+    using (var stream = fileInfo.OpenText())
+    {
+        var storedData = await stream.ReadToEndAsync();
+        grainState.State = JsonConvert.DeserializeObject(storedData, _jsonSettings);
+    }
+}
+```
+
+Note that for the deserialization, we use the `_jsonSettings` which was set on the `Init` function. This is important to be able to serialize/deserialize properly the state.
+
+## Writing State
+
+Writing the state is similar to reading the state.
+
+```csharp
+public async Task WriteStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+{
+    var storedData = JsonConvert.SerializeObject(grainState.State, _jsonSettings);
+
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+
+    using (var stream = new StreamWriter(fileInfo.Open(FileMode.Create, FileAccess.Write)))
+    {
+        await stream.WriteAsync(storedData);
+    }
+}
+```
+
+Similarly as reading, we use `_jsonSettings` to write the state.
+
+## Clearing State
+
+Clearing the state would be deleting the file if the file exists.
+
+```csharp
+public Task ClearStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+{
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+    if (fileInfo.Exists)
+        fileInfo.Delete();
```

Comment:
Probably want to set state to default. Something like:
grainState.State = Activator.CreateInstance(grainState.State.GetType())

Reply:
Nice, I added `grainState.State = Activator.CreateInstance(grainState.State.GetType());` to set the state back to a `default` instantiation.